src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/OAuth2/OAuthClient.cs

```diff
@@ -210,5 +194,111 @@ private static string CheckForExtraQueryParameter(string url)
 
             return url;
         }
+
+        public async Task<IHttpWebResponse> ExecuteRequestAsync<T>()
+        {
+            bool addCorrelationId = _requestContext != null && _requestContext.Logger.CorrelationId != Guid.Empty;
+            if (addCorrelationId)
+            {
+                _headers.Add(OAuthHeader.CorrelationId, _requestContext.Logger.CorrelationId.ToString());
+                _headers.Add(OAuthHeader.RequestCorrelationIdInResponse, "true");
+            }
+
+            HttpResponse responseMessage;
+            try
+            {               
+                if (BodyParameters != null)
+                {
+                    responseMessage = await _httpManager.SendPostAsync(
+                        new Uri(RequestUri),
+                        _headers,
+                        (Dictionary<string, string>)BodyParameters,
+                        _requestContext)
+                        .ConfigureAwait(false);
+                }
+                else
+                {
+                    responseMessage = await _httpManager.SendGetAsync(
+                         new Uri(RequestUri),
+                         _headers,
+                         _requestContext)
+                         .ConfigureAwait(false);
+                }
+            }
+            catch(TaskCanceledException ex)
+            {
+                throw new HttpRequestWrapperException(null, ex);
+            }
+
+            var webResponse = CreateResponse(responseMessage);
+
+            if (responseMessage.StatusCode != HttpStatusCode.OK)
+            {
+                throw new HttpRequestWrapperException(
+                    webResponse,
+                    new HttpRequestException(
+                        string.Format(
+                            CultureInfo.CurrentCulture,
+                            "Response status code does not indicate success: {0} ({1}).",
+                            (int)webResponse.StatusCode,
+                            webResponse.StatusCode),
+                        new AdalException(webResponse.Body)));
+            }
+
+            if (addCorrelationId)
+            {
+                VerifyCorrelationIdHeaderInResponse(webResponse.Headers, _requestContext);
+            }
+
+            return webResponse;
+        }
+
+        public static IHttpWebResponse CreateResponse(HttpResponse response)
```

Comment:
Please try to lower visibility of methods. Too many public methods - are they really needed?